<h1>Pre-Lec</h1>

## 1.

In [3]:
import pandas as pd
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv"
df = pd.read_csv(url, sep=';')
df.isna().sum()


fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

## 2.

In [5]:
rows, columns = df.shape
print(f'Number of rows: {rows}')
print(f'Number of columns: {columns}')

Number of rows: 1599
Number of columns: 12


Observations are the rows in the data set
Variables are the columns of the data set

## 3.

In [8]:
summary = df.describe()

print(summary)

       fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000         

## 4.

In [9]:
print(df.dtypes)


fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object


In [12]:
url2 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

df2 = pd.read_csv(url2)

print(df2.dtypes)
print(df2.describe())
print(df2.shape)

survived         int64
pclass           int64
sex             object
age            float64
sibsp            int64
parch            int64
fare           float64
embarked        object
class           object
who             object
adult_male        bool
deck            object
embark_town     object
alive           object
alone             bool
dtype: object
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000

A) df2.shape reports the number of rows and columns in the data set. In this case, 891 rows and 15 columns. df2.describe() only works through columns with numerical values. As you can from the df2.dtypes list, only categories with either float64 or int64 are shown in df2.describe()

B) df2.shape 891 rows because there are 891 rows. "count" in df2.describe() shows the number of non-null values. This is why the numbers provided from each shape and describe may be different. It means there are missing values for "age"

## 5. 

An attribute shows the data or properties of the dataframe. They do not need to access the actual information such as numerical values or categories etc. Instead they give you the data about the structure of the object. In the case of df.shape, it returns the number of rows and columns.

A method is an action that is performed on the object itself. They do need to access the information and in the case of df.describe(), they do.

<h1>Post-Lec</h1>

## 6.

Count: Shows the number of non-null values per column. It should be the total number of rows minus the amount of rows that contain missing values for that column.

Mean: The average of all the values in that column.

Std: The standard deviation of the values in that column. This shows how far apart the values are from the mean. A lower std means the values are closer to the mean and a higher std means the values are farther from the mean (on average).

Min: Shows the smallest value in the column

25%: Shows the 25th percentile of the values in the column. So that means 25% of the values in the column are below the given value. For example if 100 students took an exam out of 100 points. If the 25% is 50, it means 25% of the students failed the exam.

50%: Shows the 50th percentile of the values in the column.

75: Shows the 75th percentile of the values in the column.

Max: Shows the highest value in that column


## ChatBot History:

https://chatgpt.com/share/66e3639a-ca24-8009-90c6-8131c320f98b

## 7.

1.

Suppose you want to analyze data about students test and assignment grades but you only want to analyze students who have finished all their tests and assignments. So dropna would be better in this case because it removed all rows (students) who have yet to complete their tests and assignments. deleting a column wouldn't be helpful here because you may be removing data for students who HAVE completed all their tests and assignments (which you want to analyze)

2.

Suppose you want to analyze the students test and assignment grades like above. But the dataset includes student numbers and names which are irrelevant to you because you just want to analyze the students as a group instead of individually. So you can deleted the variables "name" and "id" to get rid of useless data.

3.

Because perhaps the missing data that dropna would've gotten rid of was useless data that you don't need. It would then get rid of the entire observation. That is why deleting redundant data could be important because then you can focus only on what you need to analyze.

4.

In [2]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

df = pd.read_csv(url)

df.dropna()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1           1       1  female  38.0      1      0  71.2833        C  First   
3           1       1  female  35.0      1      0  53.1000        S  First   
6           0       1    male  54.0      0      0  51.8625        S  First   
10          1       3  female   4.0      1      1  16.7000        S  Third   
11          1       1  female  58.0      0      0  26.5500        S  First   
..        ...     ...     ...   ...    ...    ...      ...      ...    ...   
871         1       1  female  47.0      1      1  52.5542        S  First   
872         0       1    male  33.0      0      0   5.0000        S  First   
879         1       1  female  56.0      0      1  83.1583        C  First   
887         1       1  female  19.0      0      0  30.0000        S  First   
889         1       1    male  26.0      0      0  30.0000        C  First   

       who  adult_male deck  embark_town alive  alone  
1    woman       False    C    Cherbourg   yes  False  
3    woman       False    C  Southampton   yes  False  
6      man        True    E  Southampton    no   True  
10   child       False    G  Southampton   yes  False  
11   woman       False    C  Southampton   yes   True  
..     ...         ...  ...          ...   ...    ...  
871  woman       False    D  Southampton   yes  False  
872    man        True    B  Southampton    no   True  
879  woman       False    C    Cherbourg   yes  False  
887  woman       False    B  Southampton   yes   True  
889    man        True    C    Cherbourg   yes   True  

[182 rows x 15 columns]

By using the dropna method, you only consider non-null values, since you do not want to analyze incomplete data in your data analysis. This will prevent several problems, such as those missing values that might spoil the entire consistency of data, which is very important for correct statistical results.

## 8.

In [27]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

df = pd.read_csv(url)


df.columns


Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

1.

df.groupby("col1")["col2"].describe() takes the values of col1 and turns them into one group per value. Then ["col2"] looks through the data in col2 (you're basically notifying the pandas that you're interested int this column and that you're about to perform an action based on that column. Then .describe() performs the action on col2 per group of col1. So the .describe() method runs on col2 for each col1 group displays that data. 

In [28]:
df.groupby("survived")["sex"].describe()

count unique     top freq
survived                          
0          549      2    male  468
1          342      2  female  233

survived = 0 means did not survive and it shows 549. Unique means there are 2 unique categories for sex which are male and female. Top shows male appeared most times and freq shows the frequency of male which was 468.

survived = 1 means survived and it shows 342. It consists of mostly females and sits at 233 females.

2.

In [29]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [30]:
df.groupby("survived")["age"].describe()

count       mean        std   min   25%   50%   75%   max
survived                                                           
0         424.0  30.626179  14.172110  1.00  21.0  28.0  39.0  74.0
1         290.0  28.343690  14.950952  0.42  19.0  28.0  36.0  80.0

From what I see, the values of count are just separated into different groups but if you add the totals up per group, it should be the same as the "count" from df.describe()

3.

A)

In [1]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

df = pd.read_csv(url)

NameError: name 'pd' is not defined

The chatbot gave a solution immediately, pointing out I had not imported pandas as pd and providing the code to fix it.

In [4]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

df = pd.read_csv(url)

B)

In [8]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)

HTTPError: HTTP Error 404: Not Found

The chatbot immediately pointed out that there was a typo in the url. It corrected me on that and provided the code to fix it (it just fixed the typo)

C)

In [9]:
df2.groupby("pclassx")["col2"].describe()

NameError: name 'df2' is not defined

ChatBot told me I need to define df2 before using it. Didn't even need to use google. But at the same time I didn't really need to use the chat either because the error itself explains 'name 'df2' is not defined'

D)

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

df = pd.read_csv(url

SyntaxError: incomplete input (2044050677.py, line 5)

Chat said to make sure all parentheses are closed. couldnt find an answer on google. I'm guessing because its such a simple mistake that people usually don't make. Google ended up showing a bunch pof different coding problems.

E)

In [3]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

df = pd.read_csv(url)

In [4]:
df.group_by("pclass")["sex"].describle()

AttributeError: 'DataFrame' object has no attribute 'group_by'

Chat had the answer pretty quickly and even told me which parts of the code is done correctly so I don't mistakenly change the correct parts. Chat pointed out both typos I had. I couldn't find the result on google because its such a specific error.

F)

In [8]:
df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

Chat said to check the column names and then make sure they are spelled correctly whereas google did not show any solutions. Instead, it showed threads where people say "KeyError occurs even though column exists"

G)

In [9]:
df.groupby(sex)["age"].describe()

NameError: name 'sex' is not defined

Chat once again gave the answer quickly and concisely whereas google could not show me a thread with the answer

## 9.

Somewhat

## Chat Link

https://chatgpt.com/share/66edfc95-b8a0-8009-a074-419d320bc629